src/dtls13.c

```diff
@@ -0,0 +1,996 @@
+    ret = dtls13_encrypt_decrypt_record_number(ssl, seq_num, seq_len,
+                                               (byte *)input + length,
+                                               DEPROTECT);
+    if (ret != 0)
+        return ret;
+
+    *header_length = length;
+
+    if (seq_len == SEQ_16_LEN) {
+        *seq_hi_present =  1;
+        *seq_hi = seq_num[0];
+        *seq_lo = seq_num[1];
+    } else {
+        *seq_hi_present = 0;
+        *seq_hi = seq_num[0];
+    }
+
+    return 0;
+}
+
+/**
+ * dtls13_handshake_recv() - process an handshake message. Deal with
+ fragmentation if needed
+ * @ssl: [in] ssl object
+ * @input: [in] input buffer
+ * @size: [in] input buffer size
+ * @type: [out] content type
+ * @processed_size: [out] amount of byte processed
+ *
+ * returns 0 on success
+ */
+int dtls13_handshake_recv(WOLFSSL *ssl, byte *input, word32 size,
+                          enum ContentType type, word32 *processed_size)
+{
+    word32 frag_off, frag_length;
+    word32 message_length;
+    byte handshake_type;
+    word32 idx;
+    int ret;
+
+    if (type != handshake)
+        return WOLFSSL_NOT_IMPLEMENTED;
+
+    idx = 0;
+    ret = GetDtlsHandShakeHeader(ssl, input, &idx, &handshake_type,
+                                 &message_length,
+                                 &frag_off, &frag_length, size);
+    if (ret != 0)
+        return PARSE_ERROR;
+
+    if (idx + frag_length > size) {
+        WOLFSSL_ERROR(INCOMPLETE_DATA);
+        return INCOMPLETE_DATA;
+    }
+
+    if (frag_off + frag_length > message_length)
+        return BUFFER_ERROR;
+
+    if (frag_off != 0 || frag_length < message_length) {
+        DtlsMsgStore(ssl, ssl->keys.curEpoch,
+                     ssl->keys.dtls_peer_handshake_number,
+                     input + DTLS_HANDSHAKE_HEADER_SZ,
+                     message_length, handshake_type,
+                     frag_off, frag_length, ssl->heap);
+
+        *processed_size = idx + frag_length;
+
+        *processed_size += ssl->keys.padSz;
+
+        if (dtls13_next_message_complete(ssl))
+            return dtls13_process_buffered_messages(ssl);
+
+        return 0;
+    }
+
+    if (ssl->keys.dtls_peer_handshake_number !=
+        ssl->keys.dtls_expected_peer_handshake_number)
+        return WOLFSSL_NOT_IMPLEMENTED;
+
+    /* FIXME: client_hello should be handled differently */
+    ssl->keys.dtls_expected_peer_handshake_number++;
+
+    ret = DoTls13HandShakeMsgType(ssl, input, &idx,
+                                  handshake_type, message_length, size);
+    if (ret != 0)
+        return ret;
+
+    *processed_size = idx;
+
+    return 0;
+}
+
+/**
+ * dtls13_fragments_continue() - keep sending pending fragments
+ * @ssl: ssl object
+ */
+int dtls13_fragments_continue(WOLFSSL *ssl)
+{
+    int ret;
+
+    ret = dtls13_send_fragmented_internal(ssl);
+    if (ret != 0)
+        ssl->error = ret;
+
+    if (ret == 0)
+        ssl->keys.dtls_handshake_number++;
+
+    return ret;
+}
```

Comment:
IMO it would be more elegant to have `dtls13_fragments_continue` return the error and the higher levels should set the `ssl->error`.

Reply:
I agree thanks